### PR TITLE
BREAKING CHANGES: Molotov evolved which gives us BREAKING CHANGE: yay

### PR DESCRIPTION
**Pull request by thebruce**

### CODE REVIEW

File: __tests__/lib/molotovProviderBase.js [hunk 1]
```diff
@@ -0,0 +1,125 @@
+const MolotovProviderBase = require('../../lib/molotovProviderBase')
+
+describe('MolotovProviderBase', () => {
+  test('can construct', () => {
+    const mpb = new MolotovProviderBase('molotov', 'Supers', 'supersNameSpace');
```
Comment (elliotttf) [inline __tests__/lib/molotovProviderBase.js]: These ones up here should have `expect.assersions()` as well. Sorry 😞

File: _errors.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
+// @flow
+/* eslint-disable */
```
Comment (elliotttf) [inline _errors.js]: Unless you _really_ want to completely disable eslint you can target the offending rules at the top of the file, e.g.

```javascript
// @flow
/* eslint-disable max-len */
// Molotov Errors
module.exports = {
  // ...
};
```

File: package.json [hunk 3]
```diff
@@ -2,38 +2,51 @@
   "name": "molotov",
   "description": "Molotov: (Mix/Plug)-in loader for explosive composeability",
   "main": "molotov.js",
+  "version": "0.0.0-development",
   "directories": {
     "example": "examples"
   },
   "dependencies": {
-    "callsite": "^1.0.0",
```
Comment (elliotttf) [inline package.json]: yas

File: package.json [hunk 4]
```diff
@@ -2,38 +2,51 @@
...
-    "config": "^1.25.1",
+    "ajv": "^6.2.0",
     "fs-extra": "^4.0.1",
     "lodash": "^4.17.4",
-    "require-directory": "^2.1.1",
-    "try-require": "^1.2.1"
+    "require-directory": "^2.1.1"
```
Comment (elliotttf) [inline package.json]: Are you still using require-directory anywhere? It's not a big problem as long as it's require time and not user time.

File: package.json [hunk 5]
```diff
@@ -2,38 +2,51 @@
...
-    "precommit": "npm run lint",
+    "commitmsg": "commitlint -e $GIT_PARAMS",
+    "test": "npm run build && jest --coverage",
+    "test-ci": "npm run build && jest --coverage --runInBand",
+    "debug-test": "node --inspect --debug-brk ./node_modules/.bin/jest --env jest-environment-node-debug",
+    "precommit": "npm run build && npm run flow && npm run lint",
     "prepush": "npm test",
-    "coverage": "nyc --reporter=lcov --reporter=html check-coverage --statements 100 --lines 100 --branches 100 --functions 100",
-    "coveralls": "cat ./coverage/lcov.info | coveralls",
+    "prepublish": "npm run build",
```
Comment (elliotttf) [inline package.json]: If you're running publish on node 8 you _might_ want to make this `prepack`, but either way 👍

File: package.json [hunk 6]
```diff
@@ -53,5 +66,32 @@
   "homepage": "https://github.com/thebruce/molotov#readme",
   "engines": {
     "node": ">=4"
+  },
+  "jest": {
+    "cacheDirectory": ".tmp/jest_cache",
+    "coverageDirectory": "__coverage__",
+    "coverageReporters": [
+      "json",
+      "json-summary",
```
Comment (elliotttf) [inline package.json]: Unless you're using the json stuff somewhere or coveralls uses it I don't think you need it, lcov, text, and text-summary should be sufficient.

File: src/molotov.js [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+   *   definitions.
+   *
+   * @returns {void}
+   */
+  setCocktails(cocktails: Array<?Cocktail>): void {
+    this.cocktails = cocktails;
+  }
+  /**
+   * Gets the Cocktail array for this molotov.
+   * It's martini time!
```
Comment (elliotttf) [inline src/molotov.js]: pretend this is a martini and not champagne

File: src/molotovProviderBase.js [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+  ProviderBase,
+  supersNameSpace,
+} from './types/molotov';
+
+// User overrides are passed in targeted to a hierarchical namespace.
+const _ = require('lodash');
+const validator = require('./validateConfig');
+
+module.exports = class MolotovProviderBase<T: string> implements ProviderBase<T> {
+  target: T
```
Comment (elliotttf) [inline src/molotovProviderBase.js]: Is `T` a flow defined type? If not can it be more descriptive?

File: src/validateConfig.js [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+  MolotovError,
+} = require('./molotovError');
+
+
+const ajv = new Ajv();
+const validate = ajv.compile(schema);
+
+module.exports = function validateConfig(config: molotovConfig) {
+  const valid = validate(config);
+  if (!valid) throw new MolotovError(MOLOTOV_MALFORMED_MOLOTOV_CONFIG);
```
Comment (elliotttf) [inline src/validateConfig.js]: Does it make sense to expose the specific errors with the passed in config?

File: __tests__/lib/molotov.js [hunk 10]
```diff
@@ -117,3 +117,50 @@ describe('Molotov Constructor Errors', () => {
...
+
+  test('Molotov Mix Supers', () => {
+    expect.assertions(1);
+    const molotov = new Molotov(molotovConfig, 'testMolotovImplementer', supers, mixins, configOverrides, [cocktailClass]);
+    expect(molotov.getMolotov()).toMatchSnapshot();
+  });
+});
+
+
+describe('Molotov no overrides no cocktails', () => {
```
Comment (elliotttf) [inline __tests__/lib/molotov.js]: Code monkey like fritos...

File: __tests__/lib/molotovProviderBase.js (hunk 1 above)
Comment (elliotttf) [reply __tests__/lib/molotovProviderBase.js]: Fixed in #57

File: src/molotovProviderBase.js (hunk 8 above)
Comment (thebruce) [reply src/molotovProviderBase.js]: It is a generic. We could make it a literal maybe?

File: src/validateConfig.js (hunk 9 above)
Comment (thebruce) [reply src/validateConfig.js]: yes. i think so. would you mind?

Comment (thebruce) [reply src/validateConfig.js]: 👍

File: src/molotovProviderBase.js (hunk 8 above)
Comment (thebruce) [reply src/molotovProviderBase.js]: ("molotovPlugins" | "supersNameSpace") 👍

File: package.json (hunk 6 above)
Comment (thebruce) [reply package.json]: No, just what is needed in coveralls i think. 👍

File: src/molotovProviderBase.js (hunk 8 above)
Comment (elliotttf) [reply src/molotovProviderBase.js]: Oh ok, I just don't know flow that well. I think it's fine 😄

File: package.json (hunk 5 above)
Comment (thebruce) [reply package.json]: I think that's a good idea. Might need to make sure the .travis.yml is up to speed 👍

File: package.json (hunk 4 above)
Comment (thebruce) [reply package.json]: nope, burn it to the ground 👍

File: package.json (hunk 3 above)
Comment (thebruce) [reply package.json]: So fab. Could i in any way ever be so fab? no probably not.

File: _errors.js (hunk 2 above)
Comment (thebruce) [reply _errors.js]: 👍

File: __tests__/lib/mixinPluginMaker.js [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+  test('Mixin Plugin Maker', () => {
+    expect.assertions(1);
+    expect(mixinPluginMaker(molotovConfig.testMolotovImplementer.molotovPlugins, mixins, supers)).toMatchSnapshot();
+  });
+
+
+  test('Supers in pluginDef that is not in supers classes', () => {
+    expect.assertions(1);
+    const molotovConfig1 = _.cloneDeep(molotovConfig);
+    molotovConfig1.testMolotovImplementer.molotovPlugins.superNameSpaceWithouCorrespondingSuperClass = molotovConfig1.testMolotovImplementer.molotovPlugins.testSuper;
```
[the remark this answers is not in the document]
Comment (thebruce) [reply __tests__/lib/mixinPluginMaker.js]: yeah, trying to be instructive with the naming makes it very long. happy to have it read easier. 👍

**Pull request merged**